web/__init__.py

```diff
@@ -34,30 +64,58 @@
 
 @app.before_first_request
 def startup():
+    """
+    Registers required services on server start
+    """
     app.svc.register("pppp", web.service.pppp.PPPPService())
     app.svc.register("videoqueue", web.service.video.VideoQueue())
     app.svc.register("mqttqueue", web.service.mqtt.MqttQueue())
     app.svc.register("filetransfer", web.service.filetransfer.FileTransferService())
 
 
+def restart():
+    """
+    Restarts the registered services to apply new changes
+    """
+    try:
+        for name in app.svc:
+            app.svc.svcs[name].restart()
+            time.sleep(2)
```

Comment:
Wow, why are we doing a sleep between each service restart?

If this is needed, there's a bug we need to fix.

Reply:
In my testing, the timing of the different services restarting was causing this function to exit before the services had stopped and restarted successfully thus reloaded to a page where there were no running services resulting in a 500 error.

Reply:
can you try this pattern:

```python
   for name in app.svc:
       app.svc.svcs[name].restart()

   for name in app.svc:
       app.svc.svcs[name].await_ready()
```

Reply:
Ok, so I tried that and it got rid of the 500 exception I was seeing previously; however, it resulted in the site redirecting and the video feed to be broken for several seconds. 

My mechanism for testing this was: 
1. Open browser to http://127.0.0.1:4470
1. Navigate to **Setup** tab
1. Click on the `Restart Services` button

<details><summary>Expand log</summary>

```py
❯ python ./ankerctl.py webserver run
 * Serving Flask app 'web'
 * Debug mode: off
[*] WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://127.0.0.1:4470
[*] Press CTRL+C to quit
[*] 127.0.0.1 - - [10/May/2023 14:01:31] "GET / HTTP/1.1" 200 -
[*] 127.0.0.1 - - [10/May/2023 14:01:31] "GET /static/vendor/bootstrap-5.3.0-alpha3-dist/css/bootstrap.min.css HTTP/1.1" 200 -
[*] 127.0.0.1 - - [10/May/2023 14:01:31] "GET /static/vendor/bootstrap-icons-1.10.5/font/bootstrap-icons.min.css HTTP/1.1" 200 -
[*] 127.0.0.1 - - [10/May/2023 14:01:31] "GET /static/ankersrv.css HTTP/1.1" 200 -
[*] 127.0.0.1 - - [10/May/2023 14:01:31] "GET /static/vendor/bootstrap-5.3.0-alpha3-dist/js/bootstrap.bundle.min.js HTTP/1.1" 200 -
[*] 127.0.0.1 - - [10/May/2023 14:01:31] "GET /static/img/setup/prusaslicer-2.png HTTP/1.1" 200 -
[*] 127.0.0.1 - - [10/May/2023 14:01:31] "GET /static/img/setup/prusaslicer-3.png HTTP/1.1" 200 -
[*] 127.0.0.1 - - [10/May/2023 14:01:31] "GET /static/img/setup/prusaslicer-1.png HTTP/1.1" 200 -
[*] 127.0.0.1 - - [10/May/2023 14:01:31] "GET /static/vendor/cash.min.js HTTP/1.1" 200 -
[*] 127.0.0.1 - - [10/May/2023 14:01:31] "GET /static/vendor/jmuxer.min.js HTTP/1.1" 200 -
[*] 127.0.0.1 - - [10/May/2023 14:01:31] "GET /static/ankersrv.js HTTP/1.1" 200 -
[*] 127.0.0.1 - - [10/May/2023 14:01:31] "GET /static/img/setup/prusaslicer-4.png HTTP/1.1" 200 -
[*] 127.0.0.1 - - [10/May/2023 14:01:31] "GET /static/vendor/bootstrap-icons-1.10.5/font/fonts/bootstrap-icons.woff2?1fa40e8900654d2863d011707b9fb6f2 HTTP/1.1" 200 -
[*] MqttQueue: Requesting start
[*] Connecting to make-mqtt.ankermake.com
[*] VideoQueue: Requesting start
[*] PPPPService: Requesting start
[*] Trying connect over pppp
[*] 127.0.0.1 - - [10/May/2023 14:01:31] "GET /favicon.ico HTTP/1.1" 404 -
[*] Established pppp connection
[*] PPPPService: Worker started
[*] VideoQueue: Worker started
[*] Connected to mqtt
[*] MqttQueue: Worker started
[*] TOPIC [/phone/maker/AK7ZRG0C35700749/notice]
[*] PPPPService: Requesting restart
[*] PPPPService: Requesting stop
[*] PPPPService: Worker stopped
[*] PPPPService: Requesting start
[*] Trying connect over pppp
[*] VideoQueue: Service requested restart.
[*] Established pppp connection
[*] PPPPService: Worker started
[*] VideoQueue: Requesting restart
[*] VideoQueue: Requesting stop
[E] CLOSE
[*] PPPPService: Service requested restart.
[*] PPPPService: Requesting stop
[*] VideoQueue: Worker stopped
[*] VideoQueue: Requesting start
[*] PPPPService: Requesting start
[*] PPPPService: Worker stopped
[*] Trying connect over pppp
[*] Established pppp connection
[*] PPPPService: Worker started
[*] VideoQueue: Worker started
[*] MqttQueue: Requesting restart
[*] MqttQueue: Requesting stop
[*] MqttQueue: Worker stopped
[*] MqttQueue: Requesting start
[*] Connecting to make-mqtt.ankermake.com
[*] Connected to mqtt
[*] MqttQueue: Worker started
[*] FileTransferService: Requesting restart
[*] FileTransferService: Requesting stop
[E] FileTransferService: Waiting for stopped thread
Traceback (most recent call last):
  File "/Users/billybryant/github/personal/ankermake-m5-protocol/web/__init__.py", line 85, in restart
    app.svc.svcs[name].await_ready()
  File "/Users/billybryant/github/personal/ankermake-m5-protocol/web/lib/service.py", line 206, in await_ready
    raise ServiceStoppedError(f"{self.name}: Waiting for stopped thread")
web.lib.service.ServiceStoppedError: FileTransferService: Waiting for stopped thread
[*] 127.0.0.1 - - [10/May/2023 14:01:38] "GET /api/ankerctl/server/reload HTTP/1.1" 302 -
[*] 127.0.0.1 - - [10/May/2023 14:01:38] "GET / HTTP/1.1" 200 -
[*] VideoQueue: Requesting stop
[*] PPPPService: Requesting stop
[*] VideoQueue: Worker stopped
[*] PPPPService: Worker stopped
[*] 127.0.0.1 - - [10/May/2023 14:01:38] "GET /static/vendor/bootstrap-5.3.0-alpha3-dist/css/bootstrap.min.css HTTP/1.1" 304 -
[*] 127.0.0.1 - - [10/May/2023 14:01:38] "GET /static/vendor/bootstrap-icons-1.10.5/font/bootstrap-icons.min.css HTTP/1.1" 304 -
[*] 127.0.0.1 - - [10/May/2023 14:01:38] "GET /static/img/setup/prusaslicer-1.png HTTP/1.1" 304 -
[*] 127.0.0.1 - - [10/May/2023 14:01:38] "GET /static/ankersrv.css HTTP/1.1" 304 -
[*] 127.0.0.1 - - [10/May/2023 14:01:38] "GET /static/img/setup/prusaslicer-2.png HTTP/1.1" 304 -
[*] 127.0.0.1 - - [10/May/2023 14:01:38] "GET /static/img/setup/prusaslicer-3.png HTTP/1.1" 304 -
[*] 127.0.0.1 - - [10/May/2023 14:01:38] "GET /static/img/setup/prusaslicer-4.png HTTP/1.1" 304 -
[*] 127.0.0.1 - - [10/May/2023 14:01:38] "GET /static/vendor/cash.min.js HTTP/1.1" 304 -
[*] TOPIC [/phone/maker/AK7ZRG0C35700749/notice]
[*] 127.0.0.1 - - [10/May/2023 14:01:38] "GET /static/vendor/jmuxer.min.js HTTP/1.1" 304 -
[*] MqttQueue: Requesting stop
[*] 127.0.0.1 - - [10/May/2023 14:01:38] "GET /static/vendor/bootstrap-5.3.0-alpha3-dist/js/bootstrap.bundle.min.js HTTP/1.1" 304 -
[*] 127.0.0.1 - - [10/May/2023 14:01:38] "GET /static/ankersrv.js HTTP/1.1" 304 -
[*] MqttQueue: Requesting start
[*] VideoQueue: Requesting start
[*] PPPPService: Requesting start
[*] Trying connect over pppp
[*] Established pppp connection
[*] PPPPService: Worker started
[*] VideoQueue: Worker started
[E] CLOSE
[*] PPPPService: Service requested restart.
[*] PPPPService: Worker stopped
[*] Trying connect over pppp
[*] VideoQueue: Service requested restart.
[E] CLOSE
[E] PPPPService: Failed to start worker: . Retrying in 1 second.
Traceback (most recent call last):
  File "/Users/billybryant/github/personal/ankermake-m5-protocol/web/lib/service.py", line 102, in _attempt_start
    self.worker_start()
  File "/Users/billybryant/github/personal/ankermake-m5-protocol/web/service/pppp.py", line 46, in worker_start
    api.process(msg)
  File "/Users/billybryant/github/personal/ankermake-m5-protocol/libflagship/ppppapi.py", line 300, in process
    raise ConnectionResetError
ConnectionResetError
[*] PPPPService: Requesting stop
[*] VideoQueue: Worker stopped
[*] PPPPService: Requesting start
[*] Trying connect over pppp
[E] CLOSE
[E] PPPPService: Failed to start worker: . Retrying in 1 second.
Traceback (most recent call last):
  File "/Users/billybryant/github/personal/ankermake-m5-protocol/web/lib/service.py", line 102, in _attempt_start
    self.worker_start()
  File "/Users/billybryant/github/personal/ankermake-m5-protocol/web/service/pppp.py", line 46, in worker_start
    api.process(msg)
  File "/Users/billybryant/github/personal/ankermake-m5-protocol/libflagship/ppppapi.py", line 300, in process
    raise ConnectionResetError
ConnectionResetError
[*] TOPIC [/phone/maker/AK7ZRG0C35700749/notice]
[*] Trying connect over pppp
[*] Established pppp connection
[*] PPPPService: Worker started
[*] VideoQueue: Worker started
[E] CLOSE
[*] PPPPService: Service requested restart.
[*] PPPPService: Worker stopped
[*] Trying connect over pppp
[*] TOPIC [/phone/maker/AK7ZRG0C35700749/notice]
[*] Established pppp connection
[*] PPPPService: Worker started
[*] VideoQueue: Service requested restart.
[*] PPPPService: Requesting stop
[*] VideoQueue: Worker stopped
[*] PPPPService: Requesting start
[*] VideoQueue: Worker started
[*] TOPIC [/phone/maker/AK7ZRG0C35700749/notice]
[*] TOPIC [/phone/maker/AK7ZRG0C35700749/notice]
[*] TOPIC [/phone/maker/AK7ZRG0C35700749/notice]
[*] TOPIC [/phone/maker/AK7ZRG0C35700749/notice]
[*] TOPIC [/phone/maker/AK7ZRG0C35700749/notice]
[*] TOPIC [/phone/maker/AK7ZRG0C35700749/notice]
[*] TOPIC [/phone/maker/AK7ZRG0C35700749/notice]
[*] TOPIC [/phone/maker/AK7ZRG0C35700749/notice]
```

</details>

Reply:
As long as it recovers correctly (which it seems to do), I think that's good enough for now.

I want to implement actual dependency control in the ServiceManager, which would resolve this problem.

It's complicated to track the dynamic relationships between the services, but the version we have now is better than an arbitrary sleep - which *will* occasionally fail. So let's go with what we have now :)